src/ProcessManager.php

```diff
@@ -895,25 +836,42 @@ protected function checkChangedFiles($restartSlaves = true)
     }
 
     /**
-     * Populate slave pool
+     * Populate the slave pool, or remove unwanted slaves if the slave pool has more workers than is necessary.
+     * This allows a reload to adjust the amount of workers.
      *
      * @return void
      */
     public function createSlaves()
     {
-        for ($i = 1; $i <= $this->slaveCount; $i++) {
-            $this->newSlaveInstance(self::CONTROLLER_PORT + $i);
+        $current = count($this->slaves->getByStatus(Slave::ANY));
+        $diff = $this->config->getSlaveCount() - $current;
+
+        if ($diff < 0) {
```

Comment:
Not sure this is a real scenario, but what would happen if I scale the pool to say 100 workers and quickly cycle back down and up. Would I see the port taken when- under load- slaves still handle requests during scaling down/up? Do we need a safeguard to not crash ppm with a slave exception?

Reply:
I tried to recreate this scenario, but I could not seem to force ppm to crash. I think what we need the most right now is some fuzz tests.

Though I did find another issue that is also affecting master. It seems the max-request worker restart can conflict with reloads in rare cases.

* `newSlaveInstance`: Now checks if the port is free *before* trying to bootstrap a worker, etc.
* `SlavePool->getByPort`: I don't think throwing an exception for a missing slave is good use for exceptions. If I kept it this way, my fix would have looked like:

```
try { 
    $slave = $this->slaves->getByPort($port);
} catch ($ex) {
    throw new Exception(...);
}
```

Not really a fan of that. Seems more sensible just to return `null`. Modified where this is used, to make sure exceptions are still thrown where they were.

Reply:
reverted the `getByPort` exception change for now; it would result in extra logging, and wasn't consistent with `getByConnection`. Left for later.